File: MacTemplet/Cagtegory/NSViewController+Helper.m


//
//  NSViewController+Helper.m
//  MacTemplet
//
//  Created by Bin Shang on 2019/6/19.
//  Copyright © 2019 Bin Shang. All rights reserved.
//

#import "NSViewController+Helper.h"
#import <objc/runtime.h>

//FOUNDATION_EXPORT Class _Nullable NSClassFromString(NSString *aClassName);
NSViewController * NSCtrlFromString(NSString *aControllerName){
    return [[NSClassFromString(aControllerName) alloc] init];
}


@implementation NSViewController (Helper)

- (NSWindow *)currentWindow{
    return objc_getAssociatedObject(self, _cmd);
}
- (void)setCurrentWindow:(NSWindow *)currentWindow{
    objc_setAssociatedObject(self, @selector(currentWindow), currentWindow, OBJC_ASSOCIATION_RETAIN);
}


@end

File: MacTemplet/CocoaExpand/NSViewController+Helper.swift
//
//  NSViewController+Helper.swift
//  MacTemplet
//
//  Created by Bin Shang on 2019/11/23.
//  Copyright © 2019 Bin Shang. All rights reserved.
//

import Cocoa

@objc public extension NSViewController {

    var tbView: NSTableView {
        get {
            var obj = objc_getAssociatedObject(self, RuntimeKeySelector(#function)) as? NSTableView;
            if obj == nil {
                obj = NSTableView.create(view.bounds);

                objc_setAssociatedObject(self, RuntimeKeySelector(#function), obj, .OBJC_ASSOCIATION_RETAIN_NONATOMIC);
            }
            return obj!;
        }
        set {
            objc_setAssociatedObject(self, RuntimeKeySelector(#function), newValue, .OBJC_ASSOCIATION_RETAIN_NONATOMIC);
        }
    }
    
    /// 关联NSMutableArray 数据容器
    var dataList: NSMutableArray {
        get {
            var obj = objc_getAssociatedObject(self, RuntimeKeySelector(#function)) as? NSMutableArray;
            if obj == nil {
                obj = [];
                objc_setAssociatedObject(self, RuntimeKeySelector(#function), obj, .OBJC_ASSOCIATION_RETAIN_NONATOMIC);
            }
            return obj!;
        }
        set {
            objc_setAssociatedObject(self, RuntimeKeySelector(#function), newValue, .OBJC_ASSOCIATION_RETAIN_NONATOMIC);
        }
    }
    
}
